### ISSUE TITLE
False positive for "lifetime must be declared" error in 2018 preview

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#51008]
This code:

```rust
#![feature(rust_2018_preview)]

trait A {
    
}

impl<F> A for F where F: FnOnce(&i32) {}

fn main() {}
```

[yields](https://play.rust-lang.org/?gist=a9af0fb502bc40ce8cbbfdae8e22e54a&version=nightly&mode=debug):

```
error[E0687]: lifetimes used in `fn` or `Fn` syntax must be explicitly declared using `<...>` binders
 --> src/main.rs:7:33
  |
7 | impl<F> A for F where F: FnOnce(&i32) {}
  |                                 ^ in-band lifetime definition

error: aborting due to previous error

For more information about this error, try `rustc --explain E0687`.
```

**Comment by nikomatsakis** [comment]
cc @cramertj =)

**Comment by nikomatsakis** [comment]
(I'm investigating a bit)
